USWDS - Packages: usa-icon

amyleadem opened the pull request:
## Related issue
Closes https://github.com/uswds/uswds/issues/4867

## Preview Link
Preview link: [usa-icon in Storybook](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.app.cloud.gov/preview/uswds/uswds/al-usa-icon-create-module/?path=/story/design-tokens-icons--icons)

## Problem statement
Users should be able to install the isolated `usa-icon` package and have access to all of the icon files they need to design and build with USWDS. Currently, the commands `gulp buildSprite`, `gulp buildSpriteStandalone`, and `gulp buildUSWDS` output to `./dist` and not `./packages/usa-icon/dist`, which means users need the full uswds package to access our icon library. 

## Solution
**Change destination of gulp tasks**
Updating the destination in our existing gulp tasks to `./packages/usa-icon/dist` will create and consolidate the correct icon assets inside the usa-icon package. Once created inside usa-icon, we can copy these consolidated assets into `./dist/img` (for backwards compatibility).

**Redefine and rename `gulp buildSpriteStandalone`**
Renaming `gulp buildSpriteStandalone` to `gulp buildIcons` and expanding its build to generate all icon assets in usa-icon gives us a simple way to assemble the usa-icon dist without needing to build all of USWDS.

## Tasks
- [x] The commands `gulp buildSprite` and `gulp buildUSWDS` should output the same assets to both `./dist` and `./packages/usa-icon/dist`.
    - Exception: in `gulp buildUSWDS`, non-icon assets like `hero.png` and `/favicons`were intentionally excluded from `./packages/usa-icon/dist` but will appear in `dist/img`.
- [x] The commands `gulp buildSprite` and `gulp buildUSWDS` should output the same assets that are currently built in the ./dist directory in the `develop` branch.
- [x] `gulp buildIcons` should output all icon .svg assets to both package dist and root dist
- [x] Unit tests check that the correct non-empty directories and files exist in usa-icon package dist

## Testing and review
To test:
1. Check file structure
    1. Open the `al-usa-icon-create-module` branch
    1. Run `rm -rf dist` to remove all previously built dist files.
    1. Run the commands below one at a time, running`rm -rf dist packages/usa-icon/dist` in between each to clear the built dist directories:
        - `gulp buildSprite`
        - `gulp buildIcons`
        - `gulp buildUSWDS`
     1. After each gulp command, confirm that the files inside `./packages/usa-icon/dist/img` match what is found in `./dist/img`
        - Exception: in `gulp buildUSWDS`, `hero.png` and `/favicons` were intentionally excluded from `./packages/usa-icon/dist` but will appear in `dist/img`.
     1. Additionally, after each gulp command, confirm that the files inside `./dist/img` in this branch match what is generated in `./dist/img` in the `develop` branch. (Reference screenshots included below for easy reference on file structure).
         - Exception: ~`gulp buildSpriteStandalone`~ `gulp buildIcons` will now output all icon .svg assets (See screenshots below)
1. Confirm icon appearance in Storybook
    1. Open [Storybook ](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.app.cloud.gov/preview/uswds/uswds/al-usa-icon-create-module/?path=/story/design-tokens-icons--icons) and confirm that icons appear as expected.
1. Run npm test
     1. Run `npm run test` from both the root directory and the `./packages/usa-icon` directory and confirm that there are no errors. 
     
## Reference
### Output to `./dist`
- `gulp buildSprite` (no change)
    
    **Before:**
    ![image](https://user-images.githubusercontent.com/93996430/180801195-c5267a00-ab9d-4385-86d9-044a0a660567.png)

     **After:**
    ![image](https://user-images.githubusercontent.com/93996430/181123710-695e1ba3-f226-4b20-afc9-5c303b889253.png)

- `gulp buildSpriteStandalone`/`gulp buildIcons` (Now. more icon assets are generated. The results are the same as `gulpbuildUSWDS` but without favicons or hero.png)
    
    **Before (`gulp buildSpriteStandalone`):**
    ![image](https://user-images.githubusercontent.com/93996430/180800705-063634f8-db42-4487-b0fa-4f44c1d435b0.png)
    
     **After (`gulp buildIcons`):**
    ![image](https://user-images.githubusercontent.com/93996430/181624548-7a180447-87b3-4ba4-b323-8fba18d1b351.png)

- `gulp buildUSWDS` (No change)

    **Before:**
    ![image](https://user-images.githubusercontent.com/93996430/180801687-7f0cd835-cfbc-44ed-81ac-37ea0e26870b.png)
    
     **After:**
    ![image](https://user-images.githubusercontent.com/93996430/181625074-53f85932-d388-4acf-bcc5-01012749fc0b.png)

---

Before opening this PR, make sure you’ve done whichever of these applies to you:

- [ ] Confirm that this code follows the [18F Front End Coding Style Guide](https://pages.18f.gov/frontend/) and [Accessibility Guide](https://pages.18f.gov/accessibility/checklist/).
- [x] Run `git pull origin [base branch]` to pull in the most recent updates from your base and check for merge conflicts. (Often, the base branch is `develop`).
- [ ] Run `npm run prettier:scss` to format code.
- [x] Run `npm test` and confirm that all tests pass.
- [ ] Run your code through [HTML_CodeSniffer](http://squizlabs.github.io/HTML_CodeSniffer/) and make sure it’s error free.

amyleadem commented on package.json:
```diff
@@ -59,7 +59,7 @@
     "federalist": "npm run build && npm run build:storybook",
     "start": "npm run start:storybook",
-    "test": "snyk test && npm run lint && gulp typecheck && gulp lintSass && gulp test",
+    "test": "snyk test && npm run lint && gulp buildUSWDS && gulp typecheck && gulp lintSass && gulp test",
```
Added `buildUSWDS` after encountering errors (for more than just usa-icon) when no dist directories existed. 

I was conflicted on adding it, so let me know if you would like to see this removed.

amyleadem commented on tasks/svg-sprite.js:
```diff
@@ -32,52 +33,57 @@ const config = {
   renameSprite,
-  cleanSprite
+  cleanSprite,
+  copyIconAssets
```
As the last task of the buildSprite tasks, copy the img assets inside usa-icon to the root dist directory. 

This line can be removed if/when we stop distributing icons to root dist/img.

mejiaj replied:
Moving usa-icon scripts to that directory would increase clarity.

amyleadem replied:
Agreed. Moved the package-specific scripts in [180b36b](https://github.com/uswds/uswds/pull/4878/commits/180b36b237b9ea8e36b42c4f3b8e05bad341759d)

amyleadem commented on tasks/copy.js:
```diff
@@ -19,14 +19,34 @@ module.exports = {
+    dutil.logMessage("copyPackageIcons", "Copying USWDS Icons to packages/usa-icon/dist/img");
+    return src([
+      "packages/**/src/img/**/*{png,svg,gif}",
+      // exclude hero images and favicons
```
I imagined these exclusions are only temporary until we start distributing these files inside their own packages.

mejiaj replied:
Wouldn't we want both? 

1. Package for single use
2. Dist for supporting current users

This task might fit better in its respective package.

amyleadem replied:
Makes sense to me! Thanks for the suggestion. Moved the scripts to usa-icon in [05439c6](https://github.com/uswds/uswds/pull/4878/commits/05439c6ee10ed47fff7c260b326a984c0eeab132)

mejiaj commented on packages/usa-icon/package.json:
```diff
@@ -0,0 +1,29 @@
+  "description": "USWDS icon library",
+  "main": "_index.scss",
+  "scripts": {
+    "test:js": "gulp buildSpriteStandalone && mocha './src/test/*.spec.js'",
```
You can add a `pretest` hook to run a task before testing.

```suggestion
    "pretest": "gulp buildSpriteStandalone",
    "test:js": "mocha './src/test/*.spec.js'",
    "test": "npm run test:js"
```

mejiaj commented on packages/usa-icon/package.json:
```diff
@@ -0,0 +1,29 @@
+    "url": "https://github.com/uswds/uswds/issues"
+  },
+  "homepage": "https://github.com/uswds/uswds#readme",
+  "dependencies": {
+    "@uswds/uswds-core": "^0.0.1-alpha.0"
+  }
```
Do you think we should include gulp and any script dependencies as devDeps here?

amyleadem replied:
Good catch. Added gulp devDependencies in [982043b](https://github.com/uswds/uswds/pull/4878/commits/982043b44257526c28f93978de3caaa3d68f6546)

mejiaj commented on packages/usa-icon/src/test/icon.spec.js:
```diff
@@ -0,0 +1,44 @@
+const assert = require("assert");
```
Is assert needed as a dev dependency?

amyleadem replied:
Added only gulp devDependencies to the package since the other dependencies (including assert) were included with node.

mejiaj commented on packages/usa-icon/src/test/icon.spec.js:
```diff
@@ -0,0 +1,44 @@
+const fs = require("fs");
+const path = require("path");
+
+const distPath = path.resolve(path.join(__dirname, "../../dist/img"));
```
Mixed thoughts here. This can be good to test in the package, but we could also move this dist check in root.

Do you see any issues when testing the `dist` on `@uswds/uswds` releases?

amyleadem replied:
The package test checks the package dist for file/directory confirmation. If we want to have a test that checks the root dist, we can add that to the root tests at a later phase.

mejiaj reviewed (changes requested):
Works well! Added some suggestions for improving code clarity.

Overall we could
- Move usa-icon exclusive tasks to the component package
- Create exports for usa-icon so we can pull assets
- Add instructions for how to use with single and bundle packages
- Add a README to describe this package
  - What does it do?
  - Where can I find documentation?
  - How does it work?

amyleadem commented:
Thanks for the comments, @mejiaj 

I have made the following updates:
- Created a package README with the uswds-fonts README as a starting point [ea34758](https://github.com/uswds/uswds/pull/4878/commits/ea34758c16655160f7b5f2fc4592ef32d134b7fe).
- Added exports to usa-icon package.json ([18e08f0](https://github.com/uswds/uswds/pull/4878/commits/18e08f0763d51d6ed1c24c89f4565975ee649736)).
- Moved the usa-icon specific scripts into svg-sprite.js ([05439c6](https://github.com/uswds/uswds/pull/4878/commits/05439c6ee10ed47fff7c260b326a984c0eeab132)), then moved svg-sprite.js to the usa-icon package ([180b36b](https://github.com/uswds/uswds/pull/4878/commits/180b36b237b9ea8e36b42c4f3b8e05bad341759d)).
- Renamed `buildSpriteStandalone` to `buildIcons` because it felt like a more accurate description of the task [4683ecb](https://github.com/uswds/uswds/pull/4878/commits/4683ecb00a2a8ec2dfbcaabd4a4fd9d41accd16d).

amyleadem commented:
Updated the exports in usa-icon/package.json in [6a7dd08](https://github.com/uswds/uswds/pull/4878/commits/6a7dd08dd0f3b8b7305d9a6e6e78c775cbb592a4). 
The following now works in the test react repo:
```
import "@uswds/usa-icon/sass";
import sprite from "@uswds/usa-icon/sprite";
import rootIcon from "@uswds/usa-icon/file.svg";
import nestedIcon from "@uswds/usa-icon/usa-icons/alarm.svg";

function UsaIcon() {
  return(
    <>
      <img
        className="usa-icon usa-icon--size-9"
        alt="Test nested icon"
        src={`${nestedIcon}`}></img>

      <img
        className="usa-icon usa-icon--size-9"
        alt="Test root icon"
        src={`${rootIcon}`}></img>

      <svg className="usa-icon usa-icon--size-9" aria-hidden="true" focusable="false" role="img">
        <use xlinkHref={`${sprite}#rain`}></use>
      </svg>
    </>
  )
}

export default UsaIcon;
```

amyleadem commented on packages/usa-icon/package.json:
```diff
@@ -0,0 +1,29 @@
+  "name": "@uswds/usa-icon",
+  "version": "0.0.1-alpha.0",
+  "description": "USWDS icon library",
+  "main": "_index.scss",
```
Updated in [6a7dd08](https://github.com/uswds/uswds/pull/4878/commits/6a7dd08dd0f3b8b7305d9a6e6e78c775cbb592a4)

mejiaj commented on packages/usa-icon/package.json:
```diff
@@ -0,0 +1,40 @@
+  "version": "0.0.1-alpha.0",
+  "description": "USWDS icon library",
+  "main": "_index.scss",
+  "scripts": {
+    "pretest": "gulp buildIcons",
+    "test:js": "mocha './src/test/*.spec.js'",
+    "test": "npm run test:js"
+  },
```
As of now, there's currently no way to generate the sprite from this package directory right?

mejiaj reviewed (commented):
Added some comments, happy to link up and discuss further if needed.
